packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp

```diff
@@ -0,0 +1,144 @@
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_MATRIXFREETENTATIVEP_KOKKOS_DEF_HPP
+#define MUELU_MATRIXFREETENTATIVEP_KOKKOS_DEF_HPP
+
+#ifdef HAVE_MUELU_KOKKOS_REFACTOR
+
+#include "MueLu_MatrixFreeTentativeP_kokkos_decl.hpp"
+
+#include "MueLu_Aggregates_kokkos.hpp"
+#include "MueLu_AmalgamationFactory_kokkos.hpp"
+#include "MueLu_AmalgamationInfo_kokkos.hpp"
+#include "MueLu_CoarseMapFactory_kokkos.hpp"
+#include "MueLu_NullspaceFactory_kokkos.hpp"
+#include "MueLu_Level.hpp"
+#include "MueLu_MasterList.hpp"
+#include "MueLu_PerfUtils.hpp"
+#include "MueLu_PFactory.hpp"
+#include "MueLu_Monitor.hpp"
+#include "MueLu_Utilities_kokkos.hpp"
+
+#include <KokkosCompat_ClassicNodeAPI_Wrapper.hpp>
+
+#include "Teuchos_ScalarTraits.hpp"
+
+namespace MueLu {
+
+  // compute Y = alpha*R*X + beta*Y
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class DeviceType>
+  void MatrixFreeTentativeP_kokkos<Scalar,LocalOrdinal,GlobalOrdinal,Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType>>::apply(const MultiVector &X, 
+                                          MultiVector &Y, 
+                                          Teuchos::ETransp mode, 
+                                          Scalar alpha, 
+                                          Scalar beta) const {
+
+    // Step 1: Y*=beta*Y, setup necessary structures
+    Y.scale(beta);
+
+    auto kokkos_view_X = X.getDeviceLocalView(Xpetra::Access::ReadOnly);
+    auto kokkos_view_Y = Y.getDeviceLocalView(Xpetra::Access::ReadWrite);
+    std::cout << "kokkos_view_X.extent(0) = " << kokkos_view_X.extent(0) << std::endl;
+    std::cout << "kokkos_view_Y.extent(0) = " << kokkos_view_Y.extent(0) << std::endl;
+
+    LO numCols = kokkos_view_X.extent(1);
+
+    if(mode == Teuchos::TRANS) { // if we're in restrictor mode
+      LO numAggs = aggregates_->GetNumAggregates();
+      Kokkos::View<LO*,DeviceType> aggPtr;
+      Kokkos::View<LO*,DeviceType> aggNodes;
+      Kokkos::View<LO*,DeviceType> unaggregated;
+      typename Aggregates_kokkos::aggregates_sizes_type::const_type aggSizes = aggregates_->ComputeAggregateSizes(); // TODO: can probably remove this and just use aggPtr if needed
+      aggregates_->ComputeNodesInAggregate(aggPtr,aggNodes,unaggregated);
+
+      // Step 2: Compute Y=Y+alpha*R*X
+      // recall R*X is an average of X over aggregates
+      Kokkos::parallel_for("MueLu:MatrixFreeTentativeR_kokkos:apply", range_type(0,numAggs), 
+        KOKKOS_LAMBDA(const int aggIdx) {
+          LO aggSize = aggSizes(aggIdx);
+          // TODO: consider team policies or alternatives for agg average?
+          // TODO: how about team policies for columns of a multivector?
+          // TODO: probably smarter to sqrt the whole aggSizes once
+          for(int colIdx=0; colIdx<numCols; ++colIdx) {
```

Comment:
Do these loops have the correct nesting for memory access?

Reply:
Probably not. I think the right way to go is a `TeamPolicy` or an `MDRangePolicy`?

Reply:
This should be done. I just went for an `MDRangePolicy` based on the column-major ordering of the multivector